samples/Common/cpp/CameraHelper_cppwinrt.cpp

```diff
@@ -0,0 +1,279 @@
+                           && (0 == formatUTF8.compare(ToUpperString(MediaEncodingSubtypes::Nv12())) || 0 == formatUTF8.compare(ToUpperString(MediaEncodingSubtypes::Yuy2())) || 0 == formatUTF8.compare(ToUpperString(MediaEncodingSubtypes::Rgb32())));
+                });
+        }
+        if (compatibleFormat == sortedMediaFrameFormats.end())
+        {
+            std::cerr << "No suitable media format found on the selected source";
+            throw new winrt::hresult(MF_E_INVALIDMEDIATYPE);
+        }
+        selectedFormat = *compatibleFormat;
+        selectedFrameSource.SetFormatAsync(selectedFormat).get();
+        selectedFormat = selectedFrameSource.CurrentFormat();
+
+        std::wcout << "Attempting to set camera source to " << selectedFormat.Subtype().c_str()
+                   << " : " << std::to_wstring(selectedFormat.VideoFormat().Width()) << "x" << std::to_wstring(selectedFormat.VideoFormat().Height())
+                   << "@" << std::to_wstring(selectedFormat.FrameRate().Numerator() / selectedFormat.FrameRate().Denominator()) << L"fps" << std::endl;
+    }
+
+    // Create FrameReader with the FrameSource that we selected in the loop above.
+    m_frameReader = m_mediaCapture.CreateFrameReaderAsync(frameSourceIterator.Current().Value()).get();
+
+    // Set up a delegate to handle the frames when they are ready
+    m_frameArrivedEventToken = m_frameReader.FrameArrived({ this, &CameraHelper::FirstFrameArrivedHandler });
+
+    // Finally start the FrameReader
+    m_frameReader.StartAsync().get();
+}
+
+//
+// Dispose of camera pipeline resources
+//
+void CameraHelper::Cleanup()
+{
+    // Revoke callback, stop FrameReader and close instances
+    if (m_frameReader != nullptr)
+    {
+        m_frameReader.FrameArrived(m_frameArrivedEventToken);
+        m_frameReader.StopAsync().get();
+        m_frameReader.Close();
+        m_frameReader = nullptr;
+    }
+    if (m_mediaCapture != nullptr)
+    {
+        m_mediaCapture.Failed(m_failureEventToken);
+        m_mediaCapture.Close();
+        m_mediaCapture = nullptr;
+    }
+}
+
+//
+// Function to handle the frame when it arrives from FrameReader
+// and send it back to registered new frame handler if it is valid
+//
+void CameraHelper::FrameArrivedHandler(MediaFrameReader FrameReader, MediaFrameArrivedEventArgs)
+{
+    MediaFrameReference mediaFrame(nullptr);
+
+    // Try to get the actual Video Frame from the FrameReader
+    mediaFrame = FrameReader.TryAcquireLatestFrame();
+    if (mediaFrame != nullptr)
+    {
+        auto vmFrame = mediaFrame.VideoMediaFrame();
+        if (vmFrame != nullptr)
+        {
+            auto videoFrame = vmFrame.GetVideoFrame();
+            m_signalFrameAvailable(videoFrame);
+        }
+        mediaFrame.Close();
+    }
+}
+
+//
+// Function to handle the first frame when it arrives from FrameReader
+// and display its format then register the main handler.
```

Comment:
why do we need this approach? isn't the format information on the FrameSource accurate?

Reply:
calling MediaFrameSource::CurrentFormat() will throw if not in MediaCaptureSharingMode::ExclusiveControl


---
In reply to: [290489579](https://github.com/microsoft/WindowsVisionSkillsPreview/pull/28#discussion_r290489579) [](ancestors = 290489579)

Reply:
But I talked to Sang and Jenny, This seems like a bug.  Since you have the repro please file a bug. 
A workaround suggested by Sang was to get this information from DeviceController instead of having to wait for the first frame arrival.

---
In reply to: [290572957](https://github.com/microsoft/WindowsVisionSkillsPreview/pull/28#discussion_r290572957) [](ancestors = 290572957,290489579)

Reply:
try 
auto format = frameSource.Controller().VideoDeviceController().GetMediaStreamProperties(frameSource.Info().MediaStreamType()).as<VideoEncodingProperties>();

---
In reply to: [290858627](https://github.com/microsoft/WindowsVisionSkillsPreview/pull/28#discussion_r290858627) [](ancestors = 290858627,290572957,290489579)

Reply:
weird this does not seem to repro for me now, I imagine I did something else at the time causing an issue. Reverting this pattern.

---
In reply to: [290858627](https://github.com/microsoft/WindowsVisionSkillsPreview/pull/28#discussion_r290858627) [](ancestors = 290858627,290572957,290489579)